packages/txservice/src/dispatch.ts

```diff
@@ -0,0 +1,195 @@
+import { BigNumber } from "ethers";
+import { BaseLogger } from "pino";
+import PriorityQueue from "p-queue";
+
+import { ChainRpcProvider } from "./provider";
+import { GasPrice, WriteTransaction } from "./types";
+import { Transaction } from "./transaction";
+import { TimeoutError, TransactionReverted } from "./error";
+
+// TODO: sep file
+// Thread safe stack manager.
+class TransactionBuffer {
+  // Both of the following are functionally treated as stacks (LIFO).
+  private validated: Transaction[] = [];
+  private pending: Transaction[] = [];
+  // Should be used to access either of the stacks.
+  private readonly accessQueue: PriorityQueue = new PriorityQueue({ concurrency: 1 });
+
+  // Push a new tx to pending.
+  public async push(transaction: Transaction) {
+    // TODO: We may want to cap the pending stack here.
+    await this.accessQueue.add(() => {
+      this.pending.push(transaction);
+    });
+  }
+
+  // Move latest pending to completed.
+  // @returns boolean whether operation was successful, or if there was no pending tx to shift.
+  public async shift(id: string): Promise<boolean> {
+    // TODO: Trim the validated stack when it gets "too long" to save memory space.
+    return await this.accessQueue.add((): boolean => {
+      const transaction = this.pending.find((transaction) => transaction.id === id);
+      if (transaction == null) {
+        return false;
+      }
+      this.validated.push(transaction);
+      return true;
+    });
+  }
```

Comment:
We should be deleting completed transactions from memory rather than hanging on to them

Reply:
I 100% agree we should eventually delete them

In fact, it might only be worth hanging onto the "last validated transaction" - see above comment - or maybe even just "last validated nonce"